File: 05_arrays/05_05_rotate_array/RotateArray.java
public class RotateArray {
    public static void main(String[] args) {
        int[] nums = new int[]{1, 2, 3, 4, 5, 6, 7};
        int k = 3;
        System.out.println("Elements:-");
        print(nums);
        rotate(nums, k);
        System.out.println("Rotated by " + k + ":-");
        print(nums);
    }

    private static void rotate(int[] nums, int k) {
        int limit = k % nums.length;
        if (k < 0) {
            k = k + nums.length;
        }

        reverse(nums, 0, limit - 1);
        reverse(nums, nums.length - limit, nums.length - 1);
        reverse(nums, 0, nums.length - 1);
    }

    private static void reverse(int[] nums, int low, int high) {
        int i = low;
        int j = high;
        while (i < j) {
            swap(nums, i++, j--);
        }
    }

    private static void swap(int[] arr, int a, int b) {
        int temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }

    private static void print(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
}

File: 05_arrays/05_05_rotate_array/rotate_array.cpp
#include <bits/stdc++.h>

using namespace std;

void reverse(vector<int> &nums, int low, int high) {
    int i = low;
    int j = high;
    while (i < j)
    {
        swap(nums[i++], nums[j--]);
    }
}

void rotate(vector<int> &nums, int k)
{
    int length = nums.size();
    k = k % length;
    if (k < 0) {
        k = k + length;
    }

    reverse(nums, 0, k - 1);
    reverse(nums, length - k, length - 1);
    reverse(nums, 0, length - 1);
}

void print(vector<int>& nums) {
    for (size_t i = 0; i < nums.size(); i++)
    {
        cout << nums[i] << " ";
    }
    cout << endl;
}

int main() {
    vector<int> nums = { 1, 2, 3, 4, 5, 6, 7 };
    int k = 3;
    cout << "Elements:-" << endl;
    print(nums);
    
    rotate(nums, k);
    
    cout << "Rotated by " << k << ":-" << endl;
    print(nums);
}